신뢰구간 추정

In [1]:
import numpy as np
from scipy import stats
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
from statsmodels.stats.proportion import proportions_ztest

In [7]:
ds_example=pd.read_csv("/home/pirl/Downloads/2. Python Data Handling/EXH_QC.csv", engine="python")
df=ds_example[['Cabbage weight']].copy()
lower,upper = stats.norm.interval(0.95, loc=2.7, scale=0.397/np.sqrt(40))
#lower,upper = stats.norm.interval(0.95,loc=df.mean(),scale=stats.sem(df))
print("신뢰구간:({0},{1})".format(lower.round(2), upper.round(2)))

신뢰구간:(2.58,2.82)


p.250

In [8]:
pd.DataFrame({'Sample':[18,18,20,21,20,23,19,18,17,21,20,19,19,18,17,19]})
lower,upper = stats.normal.interval(0.95, loc = np.mean(df), scale = 3.8/np.sqrt(20))
print("신뢰구간, ({0}, {1})".format(lower.round(2), upper.round(2)))

Sample
0       18
1       18
2       20
3       21
4       20
5       23
6       19
7       18
8       17
9       21
10      20
11      19
12      19
13      18
14      17
15      19

In [14]:
lower,upper = stats.norm.interval(0.95, loc = np.mean(df), scale = 3.8/np.sqrt(20))
print("신뢰구간, ({0}, {1})".format(lower.round(2), upper.round(2)))

신뢰구간, ([17.52], [20.85])


In [16]:
df=pd.DataFrame({'Sample':[18,18,20,21,20,23,19,18,17,21,22,20,20,21,20,19,19,18,17,19]})
lower,upper = stats.norm.interval(0.95, loc = np.mean(df), scale = stats.sem(df))
print("신뢰구간, ({0}, {1})".format(lower.round(2), upper.round(2)))

신뢰구간, ([18.8], [20.2])


p.251

In [21]:
#df를 직접 입력
df=pd.DataFrame({'Sample':[54.1,53.3,56.1,55.7,54.0,54.1,54.5,57.1,55.2,53.8,54.1,54.1,56.1,55.0,55.9,56.0,54.9,54.3,53.9,55.0]})
#정규분포 구간 추정 함수 실행
lower,upper = stats.t.interval(0.95, len(df)-1,loc = np.mean(df), scale = stats.sem(df))
print("신뢰구간, ({0}, {1})".format(lower.round(2), upper.round(2)))

신뢰구간, ([54.39], [55.33])


가설검정 / 1-Sample t-test(p.253)

2-sample t-text(p.255)

In [30]:
df=pd.DataFrame({'sample':[74.5,81.2,73.8,82.0,76.3,75.7,80.2,72.6,77.9,82.8]})
t_result= stats.ttest_1samp(df,76.7)
t, p=t_result.statistic.round(3),t_result.pvalue.round(4)
print("1-Sample t-test")
print("검정통계량:{}".format(t))
print("p-value:{0}".format(p))

1-Sample t-test
검정통계량:[0.864]
p-value:[0.4098]


In [38]:
df1=pd.DataFrame({'sample':[6,5,5,4,6,7,6,4,5,6,4,5,5,6,4,8,6,5,6,7]})
df2=pd.DataFrame({'sample':[7,5,7,8,7,8,8,5,7,6,5,5,6,6,5,7,9,7,7,8]})
t_result= stats.ttest_ind(df1,df2,equal_var=True)
t, p=t_result.statistic.round(6),t_result.pvalue.round(6)
print("2-Sample t-test")
print("t검정통계량:{}".format(t))
print("p-value:{}".format(p))

2-Sample t-test
t검정통계량:[-3.122305]
p-value:[0.003424]


In [39]:
df1=pd.DataFrame({'sample':[1.883, 1.715, 1.799, 1.768, 1.711, 1.832, 1.427, 1.344]})
df2=pd.DataFrame({'sample':[1.435, 1.572, 1.486, 1.511, 1.457, 1.548, 1.404, 1.883]})
t_result= stats.ttest_ind(df1,df2)
t, p=t_result.statistic.round(3),t_result.pvalue.round(3)
print("2-Sample t-test")
print("t검정통계량:{}".format(t))
print("p-value:{}".format(p))

2-Sample t-test
t검정통계량:[1.7]
p-value:[0.111]


Paired-t test(p.257)|

In [41]:
df1=pd.DataFrame({'sample':[6,5,5,4,6,7,6,4,5,6,4,5,5,6,4,8,6,5,6,7]})
df2=pd.DataFrame({'sample':[7,5,7,8,7,8,8,5,7,6,5,5,6,6,5,7,9,7,7,8]})
t_result= stats.ttest_rel(df1,df2)
t, p=t_result.statistic.round(3),t_result.pvalue.round(3)
print("Paired t-test")
print("t검정통계량:{}".format(t))
print("p-value:{}".format(p))

Paired t-test
t검정통계량:[-4.524]
p-value:[0.]


검정통계량:-4.524
개선 전보다 개선 후가 더 나빠졌다.

실습은 개인적으로 한 번 해보기

ANOVA(p.259)

In [44]:
df1 = pd.DataFrame({'A':[892,623,721,678,723,790,720,670,690,771]})
df2 = pd.DataFrame({'B':[721,821,910,678,723,790,711,790,745,891]})
df3 = pd.DataFrame({'C':[621,915,888,721,894,834,841,912,845,889]})
f_result=stats.f_oneway(df1,df2,df3)
f,p =f_result.statistic.round(3), f_result.pvalue.round(3)
print("One-way ANOVA")
print("F검정통계량:{}".format(f))
print("p-value:{}".format(p))

One-way ANOVA
F검정통계량:[4.263]
p-value:[0.025]


1 Proportion test(p.260)

In [45]:
from statsmodels.stats.proportion import proportions_ztest
count=15
nobs=100
value=.1

stat,pval=proportions_ztest(count,nobs,value)
print("1-Proportion test")
print('p-검정통계량:{0:.3f}'.format(stat))
print("p-value:{0:0.3f}".format(pval))

1-Proportion test
p-검정통계량:1.400
p-value:0.161


In [47]:
# n=1000으로 바꿨을 때
from statsmodels.stats.proportion import proportions_ztest
count=150
nobs=1000
value=.1

stat,pval=proportions_ztest(count,nobs,value)
print("1-Proportion test")
print('p-검정통계량:{0:.3f}'.format(stat))
print("p-value:{0:0.3f}".format(pval))

1-Proportion test
p-검정통계량:4.428
p-value:0.000


2 Proportion test(p.262)

In [ ]:
count=np.array([4,1])
nobs=np.array([1000,1200])

stat,pval=proportions_ztest(count,nobs)
print("2 Proportion test")
print("p검정통계량:{0:.3f}".format(stat))
print("p-value:{0:.3f}".format(pval))

In [50]:
count=np.array([14,5])
nobs=np.array([1000,1200])

stat,pval=proportions_ztest(count,nobs)
print("2 Proportion test")
print("p검정통계량:{0:.3f}".format(stat))
print("p-value:{0:.3f}".format(pval))

2 Proportion test
p검정통계량:2.073
p-value:0.038


chi-square test(p.264)

In [53]:
df=pd.DataFrame({'HSG':[270,260,236,234],'SS':[228,285,225,262],'SPA':[277,284,231,208]})
chi,pval,dof,expected = stats.chi2_contingency(df.T)

print("chi-square test")
print("chisq:{0:0.3f}".format(chi))
print("p:{0:0.3f}".format(pval))
print("degree of freedom:{}".format(dof))
print("expected value:\n{}".format(expected.round(3)))

chi-square test
chisq:13.366
p:0.038
degree of freedom:6
expected value:
[[258.333 276.333 230.667 234.667]
 [258.333 276.333 230.667 234.667]
 [258.333 276.333 230.667 234.667]]


# 상관, 회귀분석

Correlation Analysis(p.269)

In [54]:
df1 = [1,2,3,4,4,5,6,6,7,8]
df2 = [23,29,49,64,76,87,96,97,109,119]

corr, pval = stats.pearsonr(df1,df2)
print("Correlation Analysis")
print("corr:{0:.3f}".format(corr))
print("p:{0:0.3f}".format(pval))

Correlation Analysis
corr:0.987
p:0.000


단순회귀분석(p.272)

In [55]:
MINUTES = [1,2,3,4,4,5,6,6,7,8]
UNITS = [23,29,49,64,76,87,96,97,109,119]

MINUTES = sm.add_constant(MINUTES)
model=sm.OLS(UNITS, MINUTES)
result=model.fit()

print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.975
Model:                            OLS   Adj. R-squared:                  0.972
Method:                 Least Squares   F-statistic:                     310.6
Date:                Fri, 12 Jul 2019   Prob (F-statistic):           1.10e-07
Time:                        13:52:45   Log-Likelihood:                -30.194
No. Observations:                  10   AIC:                             64.39
Df Residuals:                       8   BIC:                             64.99
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          7.4955      4.207      1.782      0.1

/home/pirl/anaconda3/lib/python3.7/site-packages/scipy/stats/stats.py:1416: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  "anyway, n=%i" % int(n))


# 데이터 핸들링

In [56]:
from numpy.random import randn
np.random.seed(1)
df=pd.DataFrame(randn(10,2),columns=['A','B'])
df

A         B
0  1.624345 -0.611756
1 -0.528172 -1.072969
2  0.865408 -2.301539
3  1.744812 -0.761207
4  0.319039 -0.249370
5  1.462108 -2.060141
6 -0.322417 -0.384054
7  1.133769 -1.099891
8 -0.172428 -0.877858
9  0.042214  0.582815

In [57]:
df['A']

0    1.624345
1   -0.528172
2    0.865408
3    1.744812
4    0.319039
5    1.462108
6   -0.322417
7    1.133769
8   -0.172428
9    0.042214
Name: A, dtype: float64

In [58]:
df[['A']]

A
0  1.624345
1 -0.528172
2  0.865408
3  1.744812
4  0.319039
5  1.462108
6 -0.322417
7  1.133769
8 -0.172428
9  0.042214

In [59]:
df[0:5]

A         B
0  1.624345 -0.611756
1 -0.528172 -1.072969
2  0.865408 -2.301539
3  1.744812 -0.761207
4  0.319039 -0.249370

In [60]:
df.loc[0:7,['A']]

A
0  1.624345
1 -0.528172
2  0.865408
3  1.744812
4  0.319039
5  1.462108
6 -0.322417
7  1.133769

In [61]:
df.loc[:,['A']]

A
0  1.624345
1 -0.528172
2  0.865408
3  1.744812
4  0.319039
5  1.462108
6 -0.322417
7  1.133769
8 -0.172428
9  0.042214

In [62]:
df.loc[3:8]

A         B
3  1.744812 -0.761207
4  0.319039 -0.249370
5  1.462108 -2.060141
6 -0.322417 -0.384054
7  1.133769 -1.099891
8 -0.172428 -0.877858

In [63]:
df.loc[3:3]

A         B
3  1.744812 -0.761207

In [66]:
df.loc()

df.A

In [68]:
df[df.A<0]

A         B
1 -0.528172 -1.072969
6 -0.322417 -0.384054
8 -0.172428 -0.877858

In [69]:
df.query('A<0')

A         B
1 -0.528172 -1.072969
6 -0.322417 -0.384054
8 -0.172428 -0.877858

In [70]:
np.random.seed(2)
df=pd.DataFrame(randn(10,2), columns=['C','D'])

In [71]:
df

C         D
0 -0.416758 -0.056267
1 -2.136196  1.640271
2 -1.793436 -0.841747
3  0.502881 -1.245288
4 -1.057952 -0.909008
5  0.551454  2.292208
6  0.041539 -1.117925
7  0.539058 -0.596160
8 -0.019130  1.175001
9 -0.747871  0.009025

In [72]:
df[['C']]

C
0 -0.416758
1 -2.136196
2 -1.793436
3  0.502881
4 -1.057952
5  0.551454
6  0.041539
7  0.539058
8 -0.019130
9 -0.747871

In [73]:
df1=pd.DataFrame(randn(2,2),columns=['a','b'])

In [75]:
df.append(df1)

/home/pirl/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:6692: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  sort=sort)


C         D         a         b
0 -0.416758 -0.056267       NaN       NaN
1 -2.136196  1.640271       NaN       NaN
2 -1.793436 -0.841747       NaN       NaN
3  0.502881 -1.245288       NaN       NaN
4 -1.057952 -0.909008       NaN       NaN
5  0.551454  2.292208       NaN       NaN
6  0.041539 -1.117925       NaN       NaN
7  0.539058 -0.596160       NaN       NaN
8 -0.019130  1.175001       NaN       NaN
9 -0.747871  0.009025       NaN       NaN
0       NaN       NaN -0.878108 -0.156434
1       NaN       NaN  0.256570 -0.988779

In [79]:
df1=pd.DataFrame(randn(4,4),['A','B','C','D'])
df2=pd.DataFrame(randn(4,4),['A','B','C','D'])
df3=pd.DataFrame(randn(4,4),['A','B','C','D'])
df1

0         1         2         3
A -0.381516  0.066489 -1.098739  1.584487
B -2.659449 -0.091453  0.695120 -2.033467
C -0.189469 -0.077219  0.824703  1.248213
D -0.403892 -1.384519  1.367235  1.217886

In [80]:
df2

0         1         2         3
A -0.462005  0.350888  0.381866  0.566275
B  0.204208  1.406696 -1.737960  1.040824
C  0.380472 -0.217135  1.173531 -2.343603
D  1.161521  0.386078 -1.133133  0.433093

In [81]:
df3

0         1         2         3
A -0.304086  2.585295  1.835333  0.440690
B -0.719254 -0.583415 -0.325050 -0.560235
C -0.902246 -0.590972 -0.276179 -0.516884
D -0.698590 -0.928892  2.550438 -1.473173

In [82]:
result=pd.concat([df1,df2,df3])
result

0         1         2         3
A -0.381516  0.066489 -1.098739  1.584487
B -2.659449 -0.091453  0.695120 -2.033467
C -0.189469 -0.077219  0.824703  1.248213
D -0.403892 -1.384519  1.367235  1.217886
A -0.462005  0.350888  0.381866  0.566275
B  0.204208  1.406696 -1.737960  1.040824
C  0.380472 -0.217135  1.173531 -2.343603
D  1.161521  0.386078 -1.133133  0.433093
A -0.304086  2.585295  1.835333  0.440690
B -0.719254 -0.583415 -0.325050 -0.560235
C -0.902246 -0.590972 -0.276179 -0.516884
D -0.698590 -0.928892  2.550438 -1.473173

In [84]:
result = pd.concat([df1,df2,df3],axis=1)

In [85]:
result

0         1         2         3         0         1         2  \
A -0.381516  0.066489 -1.098739  1.584487 -0.462005  0.350888  0.381866   
B -2.659449 -0.091453  0.695120 -2.033467  0.204208  1.406696 -1.737960   
C -0.189469 -0.077219  0.824703  1.248213  0.380472 -0.217135  1.173531   
D -0.403892 -1.384519  1.367235  1.217886  1.161521  0.386078 -1.133133   

          3         0         1         2         3  
A  0.566275 -0.304086  2.585295  1.835333  0.440690  
B  1.040824 -0.719254 -0.583415 -0.325050 -0.560235  
C -2.343603 -0.902246 -0.590972 -0.276179 -0.516884  
D  0.433093 -0.698590 -0.928892  2.550438 -1.473173

In [86]:
result = pd.concat([df1,df2,df3],axis=1,ignore_index=True)

In [87]:
result

0         1         2         3         4         5         6   \
A -0.381516  0.066489 -1.098739  1.584487 -0.462005  0.350888  0.381866   
B -2.659449 -0.091453  0.695120 -2.033467  0.204208  1.406696 -1.737960   
C -0.189469 -0.077219  0.824703  1.248213  0.380472 -0.217135  1.173531   
D -0.403892 -1.384519  1.367235  1.217886  1.161521  0.386078 -1.133133   

         7         8         9         10        11  
A  0.566275 -0.304086  2.585295  1.835333  0.440690  
B  1.040824 -0.719254 -0.583415 -0.325050 -0.560235  
C -2.343603 -0.902246 -0.590972 -0.276179 -0.516884  
D  0.433093 -0.698590 -0.928892  2.550438 -1.473173

In [88]:
result.columns=["col_"+str(i) for i in range(12)]
result

col_0     col_1     col_2     col_3     col_4     col_5     col_6  \
A -0.381516  0.066489 -1.098739  1.584487 -0.462005  0.350888  0.381866   
B -2.659449 -0.091453  0.695120 -2.033467  0.204208  1.406696 -1.737960   
C -0.189469 -0.077219  0.824703  1.248213  0.380472 -0.217135  1.173531   
D -0.403892 -1.384519  1.367235  1.217886  1.161521  0.386078 -1.133133   

      col_7     col_8     col_9    col_10    col_11  
A  0.566275 -0.304086  2.585295  1.835333  0.440690  
B  1.040824 -0.719254 -0.583415 -0.325050 -0.560235  
C -2.343603 -0.902246 -0.590972 -0.276179 -0.516884  
D  0.433093 -0.698590 -0.928892  2.550438 -1.473173

In [89]:
(result[result.col_0 > result.col_1])

col_0     col_1     col_2     col_3     col_4     col_5     col_6  \
D -0.403892 -1.384519  1.367235  1.217886  1.161521  0.386078 -1.133133   

      col_7    col_8     col_9    col_10    col_11  
D  0.433093 -0.69859 -0.928892  2.550438 -1.473173

In [90]:
result[result.col_0 > result.col_1]

col_0     col_1     col_2     col_3     col_4     col_5     col_6  \
D -0.403892 -1.384519  1.367235  1.217886  1.161521  0.386078 -1.133133   

      col_7    col_8     col_9    col_10    col_11  
D  0.433093 -0.69859 -0.928892  2.550438 -1.473173

In [91]:
result.query('col_0 > col_1')

col_0     col_1     col_2     col_3     col_4     col_5     col_6  \
D -0.403892 -1.384519  1.367235  1.217886  1.161521  0.386078 -1.133133   

      col_7    col_8     col_9    col_10    col_11  
D  0.433093 -0.69859 -0.928892  2.550438 -1.473173

In [92]:
result['col_0'] > result['col_1']

A    False
B    False
C    False
D     True
dtype: bool